File: examples/lancedataset.py
import io

import lance
from PIL import Image

root = "/data/datasets/custom/nuimages"

train = lance.dataset(f"{root}/train.lance")
val = lance.dataset(f"{root}/val.lance")

print("Train rows:", train.count_rows())
print("Val rows:", val.count_rows())

print("\nTrain schema:")
print(train.schema)


# 随机读取第 1000 个样本
index = 1000

table = train.take([index])
row = table.to_pylist()[0]

print("\nSample:")
print("label:", row["label"])

# 如果你保存的是:
# image: binary
# label: int32
encoded = row["image"]

print("encoded bytes:", len(encoded))
print("signature:", encoded[:8])

# 解码图片
image = Image.open(io.BytesIO(encoded))

print("format:", image.format)
print("size:", image.size)
print("mode:", image.mode)

image.show()

File: crates/rivet-vision/examples/lancedataset.rs
use arrow_array::{Array, BinaryArray, Int32Array, StringArray};
use image::ImageReader;
use lance::Dataset;
use lance::dataset::ProjectionRequest;
use std::io::Cursor;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let dataset = Dataset::open("/data/datasets/custom/nuimages/train.lance").await?;

    let projection =
        ProjectionRequest::from_columns(["image", "label", "label_name", "path"], dataset.schema());

    let batch = dataset.take(&[1000], projection).await?;

    let image_col = batch
        .column_by_name("image")
        .ok_or("missing image column")?
        .as_any()
        .downcast_ref::<BinaryArray>()
        .ok_or("image is not BinaryArray")?;

    let label_col = batch
        .column_by_name("label")
        .ok_or("missing label column")?
        .as_any()
        .downcast_ref::<Int32Array>()
        .ok_or("label is not Int32Array")?;

    let label_name_col = batch
        .column_by_name("label_name")
        .ok_or("missing label_name column")?
        .as_any()
        .downcast_ref::<StringArray>()
        .ok_or("label_name is not StringArray")?;

    let path_col = batch
        .column_by_name("path")
        .ok_or("missing path column")?
        .as_any()
        .downcast_ref::<StringArray>()
        .ok_or("path is not StringArray")?;

    // zero-copy borrow of encoded JPEG bytes
    let encoded: &[u8] = image_col.value(0);

    let label = label_col.value(0);
    let label_name = label_name_col.value(0);
    let path = path_col.value(0);

    println!("label: {label}");
    println!("label_name: {label_name}");
    println!("path: {path}");
    println!("encoded bytes: {}", encoded.len());
    println!("signature: {:02X?}", &encoded[..encoded.len().min(10)]);

    // JPEG decode
    let image = ImageReader::new(Cursor::new(encoded))
        .with_guessed_format()?
        .decode()?;

    let rgb = image.into_rgb8();

    let width = rgb.width();
    let height = rgb.height();

    // move out the underlying pixel Vec<u8>
    let pixels: Vec<u8> = rgb.into_raw();

    println!("decoded: {width}x{height}");
    println!("shape: [{height}, {width}, 3]");
    println!("pixel bytes: {}", pixels.len());

    Ok(())
}
